runtime/lua/lsp/client.lua

```diff
@@ -0,0 +1,342 @@
+local json = require('runtime.lua.json')
+local util = require('runtime.lua.builtin_util')
+
+local protocol = require('runtime.lua.lsp.protocol')
+local message = require('runtime.lua.lsp.message')
+local lsp_doautocmd = require('runtime.lua.lsp.autocmds').lsp_doautocmd
+local get_callback_function = require('runtime.lua.lsp.callbacks').get_callback_function
+
+local log = require('runtime.lua.lsp.lsp').log
+
+local read_state = {
+  init = 0,
+  header = 1,
+  body = 2,
+}
+
+local error_level = {
+  critical = 0,
+  reset_state = 1,
+  info = 2,
+}
+
+
+local active_jobs = {}
+
+active_jobs.add = function(id, obj)
+  active_jobs[id] = obj
+end
+
+active_jobs.remove = function(id)
+  active_jobs[id] = nil
+end
+
+local client = {}
+client.__index = client
+
+client.job_stdout = function(id, data)
+  if active_jobs[id] == nil then
+    return
+  end
+
+  active_jobs[id]:on_stdout(data)
+end
+client.new = function(name, ft, cmd, args)
+  log.trace('Starting new client: ' .. name)
+
+  -- TODO: I'm a little concerned about the milliseconds after starting up the job.
+  -- Not sure if we'll register ourselves faster than we will get stdin or out that we want...
+  local job_id = vim.api.nvim_call_function('lsp#job#start', {cmd, args})
+
+  local self = setmetatable({
+    job_id = job_id,
+
+    name = name,
+    ft = ft,
+    cmd = cmd,
+    args = args,
+
+    -- State for handling messages
+    _read_state = read_state.init,
+    _read_data = '',
+    _current_header = {},
+
+    -- Results & Callback handling
+    --  Callbacks must take two arguments:
+    --      1 - success: true if successful, false if error
+    --      2 - data: corresponding data for the message
+    _callbacks = {},
+    _results = {},
+  }, client)
+
+  active_jobs.add(job_id, self)
+
+  return self
+end
+client.initialize = function(self)
```

Comment:
Newline between functions?

Perhaps we should converge on some standard formatter so we don't get bikeshedding wars? I found two:

- https://github.com/trixnz/lua-fmt
- https://github.com/CurtisFenner/luafmt

Of which the second seems low-dependency (only Lua itself), while the first seems a bit more mature (but written in TypeScript).

The reason I suggest this is because I find gofmt very liberating when coding in Go. I haven't tried any of them, btw.

Reply:
A Lua formatter would be very welcome, I wasn't aware that any existed.